include/kernels/staggered_quark_smearing.cuh

```diff
@@ -0,0 +1,229 @@
+    static constexpr QudaGhostExchange ghost = QUDA_GHOST_EXCHANGE_PAD;
+    typedef typename gauge_mapper<Float, reconstruct, 18, QUDA_STAGGERED_PHASE_NO, gauge_direct_load, ghost>::type G;
+
+    typedef typename mapper<Float>::type real;
+
+    F out;        /** output vector field */
+    const F in;   /** input vector field */
+    const F in_pack; /** input vector field used in packing to be able to independently resetGhost */
+    const G U;    /** the gauge field */
+    int dir;      /** The direction from which to omit the derivative */
+    int t0;
+    bool is_t0_kernel;
+    int t0_offset;
+    int t0_face_offset[4];
+    int face_size[4];
+    int t0_face_size[4];
+    int threadDimMapUpper_t0[4];
+    int threadDimMapLower_t0[4];
+
+    StaggeredQSmearArg(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, int t0, bool is_t0_kernel, int parity, int dir, 
+               bool dagger, const int *comm_override) :
+
+      DslashArg<Float, nDim>(in, U, parity, dagger, is_t0_kernel, 3, false, comm_override),//FIXME reuse xpay argument for t0 option
+      out(out, 3),
+      in(in, 3),
+      in_pack(in, 3),
+      U(U),
+      dir(dir),
+      t0(t0),
+      is_t0_kernel(is_t0_kernel),
+      t0_offset(is_t0_kernel ? in.VolumeCB() / in.X(3) : 0)
+    {
+      if (in.V() == out.V()) errorQuda("Aliasing pointers");
+      checkOrder(out, in);        // check all orders match
+      checkPrecision(out, in, U); // check all precisions match
+      checkLocation(out, in, U);  // check all locations match
+      if (!in.isNative() || !U.isNative())
+        errorQuda("Unsupported field order colorspinor(in)=%d gauge=%d combination\n", in.FieldOrder(), U.FieldOrder());
+      if (dir < 3 || dir > 4) errorQuda("Unsupported laplace direction %d (must be 3 or 4)", dir);
+
+      for( int i=0; i<4; i++ )
+      {
+        t0_face_offset[i] = is_t0_kernel ? (int)(this->dc.face_XYZ[i])/2 : 0;
+        face_size[i] = 3 * this->dc.ghostFaceCB[i]; // 3=Nface
+        t0_face_size[i] = ( is_t0_kernel && i<3 ) ? face_size[i] / in.X(3) : face_size[i];
+      }
+
+      // partial replication of dslash::setFusedParam()
+      int prev = -1;
+      for( int i=0; i<4; i++ )
+      {
+        threadDimMapLower_t0[i] = 0;
+        threadDimMapUpper_t0[i] = 0;
+        if( !(this->commDim[i]) ) continue;
+        threadDimMapLower_t0[i] = ( prev>=0 ? threadDimMapUpper_t0[prev] : 0 );
+        threadDimMapUpper_t0[i] = threadDimMapLower_t0[i] + 2 * t0_face_size[i];
+        prev = i;
+      }
+    }
+  };
+
+  /**
+     Applies the off-diagonal part of the covariant derivative operator
+
+     @param[out] out The out result field
+     @param[in,out] arg Parameter struct
+     @param[in] U The gauge field
+     @param[in] coord Site coordinate struct
+     @param[in] parity The site parity
+     @param[in] idx Thread index (equal to face index for exterior kernels)
```

Comment:
Doxygen list doesn't quite match the function parameters being passed

Reply:
fixed

Reply:
This still is quite fixed (`active` is missing)